server/handlers/validation.go

```diff
@@ -192,71 +201,78 @@ func loadTargetsFromStore(gun, role string, repo *tuf.Repo, store storage.MetaSt
 	return repo.SetTargets(role, t)
 }
 
+// generateSnapshot generates a new snapshot from the previous one in the store - this assumes all
+// the other roles except timestamp have already been set on the repo, and will set the generated
+// snapshot on the repo as well
 func generateSnapshot(gun string, repo *tuf.Repo, store storage.MetaStore) (*storage.MetaUpdate, error) {
-	role, err := repo.GetBaseRole(data.CanonicalSnapshotRole)
-	if err != nil {
-		return nil, validation.ErrBadRoot{Msg: "root did not include snapshot role"}
+	var prev *data.SignedSnapshot
+	currentJSON, err := store.GetCurrent(gun, data.CanonicalSnapshotRole)
+	if err == nil {
+		prev = new(data.SignedSnapshot)
+		if err = json.Unmarshal(currentJSON, prev); err != nil {
+			logrus.Error("Failed to unmarshal existing snapshot for GUN ", gun)
+			return nil, err
+		}
 	}
 
-	algo, keyBytes, err := store.GetKey(gun, data.CanonicalSnapshotRole)
-	if err != nil {
-		return nil, validation.ErrBadHierarchy{Msg: "could not retrieve snapshot key. client must provide snapshot"}
+	if _, ok := err.(storage.ErrNotFound); !ok && err != nil {
+		return nil, validation.ErrValidation{Msg: err.Error()}
```

Comment:
Can we test this case somehow?  Maybe setting a bad snapshot with the swizzler or in `snapshot_test.go`?